### PR TITLE
Github Actions

### LINKED ISSUE #26: Improve github actions with java/ts format and lint check and sonar
Effectively:

```
# Can it be better than this with a mvn command?
- cd frontend/web && yarn format:check && yarn lint
- cd backend && mvn spotless:check
```

For sonarqube see other projects.

### PR DESCRIPTION
**Pull request by jonnyelliot**
GitHub actions for:

- Branch builds (non-main/master): check formatting, run maven package.
- Main/Master builds: check formatting, run maven verify: which is maven package + run integration tests that call out to GitHub. Builds Docker images.

All builds have caching for maven, yarn and eslint

Closes #26

### CODE REVIEW

File: .github/workflows/pr-to-main.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
+name: PR onto main
```
Comment (chrisflatley) [inline .github/workflows/pr-to-main.yml:1]: Best rename if it's an any pr build

File: .github/workflows/pr-to-main.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      with:
+        path: "**/*.eslintcache"
+        key: ${{ runner.os }}-eslintcache-${{ github.head_ref }}
+
+    # Actual process of testing
+    - name: Build with Maven
+      run: mvn -B package --file pom.xml
+
+    # Check formatting
+    - name: Check backend formatting
```
Comment (chrisflatley) [inline .github/workflows/pr-to-main.yml:59]: Move the formatting checks above the build

File: .github/workflows/quick-build.yml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      run: echo "::set-output name=dir::$(mvn help:evaluate -Dexpression=settings.localRepository -q -DforceStdout)"
+    - uses: actions/cache@v2
+      id: maven-cache
+      with:
+        path: ${{ steps.maven-cache-dir-path.outputs.dir }}
+        key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
+        restore-keys: |
+          ${{ runner.os }}-maven-
+    - name: Build with Maven
+      run: mvn -B package --file pom.xml
```
Comment (chrisflatley) [inline .github/workflows/quick-build.yml:45]: This should still do the format check. 

Prs should run the tests 

The main branch should have integration tests run too.

**Comment by jonnyelliot** [comment]
@chrisflatley I made this as a draft PR to test the actions. Wasn't expecting a review so soon! All of your comments have been addressed.

File: .github/workflows/branch.yml [hunk 4]
```diff
@@ -0,0 +1,58 @@
+name: Branch Build
```
Comment (chrisflatley) [inline .github/workflows/branch.yml:1]: DElete this

File: .github/workflows/pull-request.yml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      with:
+        path: ${{ steps.maven-cache-dir-path.outputs.dir }}
+        key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
+        restore-keys: |
+          ${{ runner.os }}-maven-
+    - uses: actions/cache@v2
+      id: eslint-cache
+      with:
+        path: "**/*.eslintcache"
+        key: ${{ runner.os }}-eslintcache-${{ github.head_ref }}
```
Comment (chrisflatley) [inline .github/workflows/pull-request.yml:50]: Possibly change to branch name?

File: .github/workflows/full.yml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      with:
+        path: ${{ steps.maven-cache-dir-path.outputs.dir }}
+        key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
+        restore-keys: |
+          ${{ runner.os }}-maven-
+    - uses: actions/cache@v2
+      id: eslint-cache
+      with:
+        path: "**/*.eslintcache"
+        key: ${{ runner.os }}-eslintcache-${{ github.head_ref }}
```
Comment (chrisflatley) [inline .github/workflows/full.yml:49]: branch name.

Possibly should delete each day - expiry?

**Review by chrisflatley — APPROVED** [review]

**Pull request merged**